### PR TITLE
FIXUP: check invalid input for dim argument (issue #1226)

### PR DESCRIPTION
**Pull request by alpinho**
Attempts to fix issue #1226 plus check whether input for dim argument is out of range. Please, review it.

### CODE REVIEW

File: nilearn/plotting/img_plotting.py [hunk 1]
```diff
@@ -383,6 +383,17 @@ def _load_anat(anat_img=MNI152TEMPLATE, dim='auto', black_bg='auto'):
             else:
                 black_bg = True
     if dim:
+        if dim != 'auto' and not isinstance(dim, numbers.Number):
+            raise ValueError(
+                "The input given for 'dim' needs to be "
+                "a float. "
+                "You provided dim=%s in %s" % (str(dim), type(dim)))
+        elif dim != 'auto' and dim not in range(-1, 1):
```
Comment (bthirion) [inline nilearn/plotting/img_plotting.py]: I disagree with that one: to render Haby's dataset anat images I had to use dim=-1.5.

File: nilearn/plotting/tests/test_img_plotting.py [hunk 2]
```diff
@@ -937,3 +937,11 @@ def test_plotting_functions_with_nans_in_bg_img():
...
 
     plt.close()
+
+
+def test_plotting_functions_with_dim_invalid_input():
+    # Test whether error raises with bad error to input
+    img = _generate_img()
+    assert_raises(ValueError, plot_roi, img, dim='-10')
+    assert_raises(ValueError, plot_stat_map, img, dim='-10')
+    assert_raises(ValueError, plot_anat, img, dim='-10')
```
Comment (bthirion) [inline nilearn/plotting/tests/test_img_plotting.py]: I think that the three are redundant: only one would be enough

File: nilearn/plotting/img_plotting.py (hunk 1 above)
Comment (KamalakerDadi) [inline nilearn/plotting/img_plotting.py]: But, the documentation says range is in (-1, 1) ?

File: nilearn/plotting/tests/test_img_plotting.py (hunk 2 above)
Comment (alpinho) [inline nilearn/plotting/tests/test_img_plotting.py]: I removed the plot_roi and plot_anat test.

**Comment by bthirion** [comment]
The documentation is reasonable, but data are not always reasonable...Could the code issue a warning rather than an error ?



Sent from my smartphone. Sorry for misspelling and brevity.

-------- Message d'origine --------
De : KamalakerDadi <notifications@github.com> 
Date : 24/06/2017  14:13  (GMT+01:00) 
À : nilearn/nilearn <nilearn@noreply.github.com> 
Cc : bthirion <bertrand.thirion@inria.fr>, Comment <comment@noreply.github.com> 
Objet : Re: [nilearn/nilearn] FIXUP: check invalid input for dim argument
  (issue #1226) (#1474) 

@KamalakerDadi commented on this pull request.



In nilearn/plotting/img_plotting.py:
> @@ -383,6 +383,17 @@ def _load_anat(anat_img=MNI152TEMPLATE, dim='auto', black_bg='auto'):
             else:
                 black_bg = True
     if dim:
+        if dim != 'auto' and not isinstance(dim, numbers.Number):
+            raise ValueError(
+                "The input given for 'dim' needs to be "
+                "a float. "
+                "You provided dim=%s in %s" % (str(dim), type(dim)))
+        elif dim != 'auto' and dim not in range(-1, 1):

But, the documentation says range is in (-1, 1) ?

—
You are receiving this because you commented.
Reply to this email directly, view it on GitHub, or mute the thread.


  
  




{"api_version":"1.0","publisher":{"api_key":"05dde50f1d1a384dd78767c55493e4bb","name":"GitHub"},"entity":{"external_key":"github/nilearn/nilearn","title":"nilearn/nilearn","subtitle":"GitHub repository","main_image_url":"https://cloud.githubusercontent.com/assets/143418/17495839/a5054eac-5d88-11e6-95fc-7290892c7bb5.png","avatar_image_url":"https://cloud.githubusercontent.com/assets/143418/15842166/7c72db34-2c0b-11e6-9aed-b52498112777.png","action":{"name":"Open in GitHub","url":"https://github.com/nilearn/nilearn"}},"updates":{"snippets":[{"icon":"PERSON","message":"@KamalakerDadi commented on #1474"}],"action":{"name":"View Pull Request","url":"https://github.com/nilearn/nilearn/pull/1474#discussion_r123876114"}}}

File: nilearn/plotting/img_plotting.py (hunk 1 above)
Comment (GaelVaroquaux) [inline nilearn/plotting/img_plotting.py]: The documentation is wrong :(.

I think that it's in [-infty, 1]

**Comment by GaelVaroquaux** [comment]
Another problem is that I think that the line of code added is wrong: 
<pre>
In [1]: .5 in range(-1, 1)
Out[1]: False
</pre>

Indeed, range(-1, 1) is a set of integers, and not a continuous range.

File: nilearn/plotting/img_plotting.py [hunk 3]
```diff
@@ -383,6 +383,17 @@ def _load_anat(anat_img=MNI152TEMPLATE, dim='auto', black_bg='auto'):
             else:
                 black_bg = True
     if dim:
+        if dim != 'auto' and not isinstance(dim, numbers.Number):
+            raise ValueError(
```
Comment (KamalakerDadi) [inline nilearn/plotting/img_plotting.py]: >Could the code issue a warning rather than an error ?

@bthirion You mean here raising a warning than a error ?

Comment (bthirion) [inline nilearn/plotting/img_plotting.py]: Yes.

Comment (KamalakerDadi) [inline nilearn/plotting/img_plotting.py]: Raise warning and override given value to 'auto'. Sounds  good ?

Comment (bthirion) [inline nilearn/plotting/img_plotting.py]: Yes, but from reading the code, I think that we can accept values between -2 and +2.

**Comment by KamalakerDadi** [comment]
From the discussions and comments on this PR. To summarize, correct me if I am wrong. Fix would be:

-  To add in the documentation at related plotting functions saying "typically it should be in between -2 and 2".
- Raise an error or warning if dim is not 'auto' or not a number. I am more leaning towards raising an error.

**Comment by bthirion** [comment]
Sounds good for me.

**Comment by GaelVaroquaux** [comment]
> Sounds good for me.

+1

**Comment by alpinho** [comment]
I went through the two bullet points mentioned by Kamalakar. Plus, I have already removed the test to two out of the three functions previously included in the function in test_img_plotting.py in order to avoid redundancy.

**Comment by KamalakerDadi** [comment]
ping @GaelVaroquaux merge ?

**Comment by GaelVaroquaux** [comment]
LGTM. Merging. Thank you

**Pull request merged**